File: server-erlang/apps/server_erlang/src/websocket_handler.erl
-module(websocket_handler).
-behaviour(cowboy_websocket).

-export([init/2, websocket_init/1, websocket_handle/2, websocket_info/2, terminate/3]).

%% Função init/2
init(Req, _Opts) ->
    {cowboy_websocket, Req, #{}}.

%% Inicialização do WebSocket
websocket_init(State) ->
    {ok, State}.

%% Tratamento de mensagens recebidas
websocket_handle({text, Msg}, State) ->
    {reply, {text, <<"Echo: ", Msg/binary>>}, State};
websocket_handle(_Frame, State) ->
    {ok, State}.

%% Tratamento de outras informações
websocket_info(_Info, State) ->
    {ok, State}.

%% Função de término
terminate(_Reason, _Req, _State) ->
    ok.

File: server_elixir/lib/server_elixir/websocket_handler.ex
defmodule ServerElixir.WebSocketHandler do
  @behaviour :cowboy_websocket

  def init(req, state) do
    {:cowboy_websocket, req, state}
  end

  def websocket_init(state) do
    ServerElixir.Clients.add(self())
    clients_size = ServerElixir.Clients.size()
    IO.puts("WebSocket connection established, #{clients_size} clients connected")

    if clients_size == 32 do
      ServerElixir.Clients.send_ready_message()
    end

    {:ok, state}
  end

  def websocket_handle({:text, message}, state) do
    ServerElixir.Clients.broadcast(self(), message)
    {:ok, state}
  end

  def websocket_info(:send_ready, state) do
    IO.puts("Sending ready message to clients")
    {:reply, {:text, "ready"}, state}
  end

  def websocket_info({:send_message, message}, state) do
    {:reply, {:text, message}, state}
  end

  def terminate(_reason, _req, _state) do
    ServerElixir.Clients.remove(self())
    IO.puts("WebSocket connection terminated")
    :ok
  end
end
